core/templates/pages/exploration-editor-page/services/exploration-save.service.spec.ts

```diff
@@ -0,0 +1,777 @@
+    spyOn(changeListService, 'discardAllChanges')
+      .and.returnValue(Promise.reject(null));
+    let modalSpy = spyOn($uibModal, 'open').and.returnValue(
+      {
+        opened: Promise.resolve(),
+        result: Promise.resolve(['1'])
+      });
+    spyOn(ExplorationRightsService, 'publish')
+      .and.resolveTo();
+
+    explorationSaveService.showPublishExplorationModal(
+      startLoadingCb, endLoadingCb);
+    $rootScope.$apply();
+    tick();
+    $rootScope.$apply();
+    tick();
+    $rootScope.$apply();
+
+    expect(modalSpy).toHaveBeenCalled();
+  }));
+
+  it('should not publish exploration in case of backend ' +
+    'error', fakeAsync(function() {
+    spyOn($uibModal, 'open').and.returnValue({
+      result: Promise.reject('failure')
+    });
+    let publishSpy = spyOn(ExplorationRightsService, 'publish')
+      .and.resolveTo();
+
+    ExplorationTitleService.savedMemento = true;
+    ExplorationObjectiveService.savedMemento = true;
+    ExplorationCategoryService.savedMemento = true;
+    explorationLanguageCodeService.savedMemento = 'afk';
+    explorationTagsService.savedMemento = 'invalid';
+
+    explorationSaveService.showPublishExplorationModal();
+    $rootScope.$apply();
+    tick();
+    $rootScope.$apply();
+    tick();
+    $rootScope.$apply();
+
+    expect(publishSpy).not.toHaveBeenCalled();
+  }));
+
+  it('should mark exploaration as editable', fakeAsync(function() {
+    let editableSpy = spyOn(editabilityService, 'markNotEditable')
+      .and.returnValue(null);
+    let startLoadingCb = jasmine.createSpy('startLoadingCb');
+    let endLoadingCb = jasmine.createSpy('endLoadingCb');
+    spyOn(changeListService, 'discardAllChanges')
+      .and.returnValue(Promise.resolve(null));
+    spyOn($uibModal, 'open').and.returnValue(
+      {
+        opened: Promise.resolve(),
+        result: Promise.resolve(['1'])
+      });
+
+    explorationSaveService.showPublishExplorationModal(
+      startLoadingCb, endLoadingCb);
+    $rootScope.$apply();
+    tick();
+    $rootScope.$apply();
+    tick();
+    $rootScope.$apply();
+
+    expect(editableSpy).toHaveBeenCalled();
+  }));
+
+  it('should open publish exploration modal', fakeAsync(function() {
+    let modalSpy = spyOn($uibModal, 'open').and.callThrough();
+    let startLoadingCb = jasmine.createSpy('startLoadingCb');
+    let endLoadingCb = jasmine.createSpy('endLoadingCb');
+
+    explorationSaveService.showPublishExplorationModal(
+      startLoadingCb, endLoadingCb);
+    $rootScope.$apply();
+    tick();
+    $rootScope.$apply();
+    tick();
+    $rootScope.$apply();
+
+    expect(modalSpy).toHaveBeenCalled();
+  }));
+
+  it('should check whether the exploration is saveable', function() {
+    spyOn(changeListService, 'isExplorationLockedForEditing')
+      .and.returnValue(false);
+
+    let result = explorationSaveService.isExplorationSaveable();
+
+    expect(result).toBe(false);
+  });
+});
+
+describe('Exploration save service ' +
```

Comment:
Is it possible to keep this describe inside the one above?  It will make the code a bit more compact.
Same for the other describe stmts

Reply:
It will throw an error.
(Can't initialize two testbed modules in the same describe block.)